apollo-router/src/plugins/traffic_shaping/retry.rs

```diff
@@ -76,3 +118,129 @@ impl<Res, E> Policy<subgraph::Request, Res, E> for RetryPolicy {
+                                .build(),
+                        )
+                        .build()
+                        .unwrap(),
+                )
+                .build();
+
+            assert!(retry
+                .retry(
+                    &subgraph_req,
+                    Err(&Box::new(FetchError::SubrequestHttpError {
+                        status_code: None,
+                        service: String::from("my_subgraph_name_error"),
+                        reason: String::from("cannot contact the subgraph"),
+                    }))
+                )
+                .is_some());
+
+            assert!(retry
+                .retry(
+                    &subgraph_req,
+                    Err(&Box::new(FetchError::SubrequestHttpError {
+                        status_code: None,
+                        service: String::from("my_subgraph_name_error"),
+                        reason: String::from("cannot contact the subgraph"),
+                    }))
+                )
+                .is_some());
+
+            assert_counter!(
+                "apollo_router_http_request_retry_total",
+                2,
+                "subgraph" = "my_subgraph_name_error"
+            );
+        }
+        .with_metrics()
+        .await;
+    }
+
+    #[tokio::test]
+    async fn test_retry_with_http_status_code() {
+        async {
+            let retry = RetryPolicy::new(
+                Some(Duration::from_secs(10)),
+                Some(10),
+                Some(0.2),
+                Some(false),
+            );
+
+            let subgraph_req = subgraph::Request::fake_builder()
+                .subgraph_name("my_subgraph_name_error")
+                .subgraph_request(
+                    http_ext::Request::fake_builder()
+                        .header("test", "my_value_set")
+                        .body(
+                            graphql::Request::fake_builder()
+                                .query(String::from("query { test }"))
+                                .build(),
+                        )
+                        .build()
+                        .unwrap(),
+                )
+                .build();
+
+            assert!(retry
+                .retry(
+                    &subgraph_req,
+                    Ok::<&subgraph::Response, &BoxError>(
+                        &subgraph::Response::fake_builder()
+                            .status_code(StatusCode::BAD_REQUEST)
+                            .build()
+                    )
+                )
+                .is_some());
+
+            assert!(retry
+                .retry(
+                    &subgraph_req,
+                    Ok::<&subgraph::Response, &BoxError>(
+                        &subgraph::Response::fake_builder()
+                            .status_code(StatusCode::BAD_REQUEST)
+                            .build()
+                    )
+                )
+                .is_some());
```

Comment:
Add a test for a redirect: [MOVED_PERMANENTLY](https://docs.rs/http/latest/http/status/struct.StatusCode.html#associatedconstant.MOVED_PERMANENTLY)